src/Logic.lidr

```diff
@@ -0,0 +1,1296 @@
+may render it _inconsistent_ -- that is, they may make it possible to prove
+every proposition, including \idr{Void}!
+
+Unfortunately, there is no simple way of telling whether an axiom is safe to
+add: hard work is generally required to establish the consistency of any
+particular combination of axioms.
+
+However, it is known that adding functional extensionality, in particular, _is_
+consistent.
+
+\todo[inline]{Is there such a command in Idris?}
+
+To check whether a particular proof relies on any additional axioms, use the
+Print Assumptions command.
+
+Print Assumptions function_equality_ex2.
+(* ===>
+     Axioms:
+     functional_extensionality :
+         forall (X Y : Type) (f g : X -> Y),
+                (forall x : X, f x = g x) -> f = g *)
+
+
+==== Exercise: 4 stars (tr_rev)
+
+One problem with the definition of the list-reversing function \idr{rev} that we
+have is that it performs a call to \idr{++} on each step; running \idr{++} takes
+time asymptotically linear in the size of the list, which means that \idr{rev}
+has quadratic running time. 
+
+> rev : (l : List x) -> List x
+> rev [] = []
+> rev (h::t) = (rev t) ++ [h]
+
+We can improve this with the following definition:
+
+> rev_append : (l1, l2 : List x) -> List x
+> rev_append [] l2 = l2
+> rev_append (x :: xs) l2 = rev_append xs (x :: l2)
+
+> tr_rev : (l : List x) -> List x
+> tr_rev l = rev_append l []
+
+(This is very similar to how \idr{reverse} is defined in \idr{Prelude.List}.)
+
+This version is said to be _tail-recursive_, because the recursive call to the
+function is the last operation that needs to be performed (i.e., we don't have
+to execute \idr{++} after the recursive call); a decent compiler will generate
+very efficient code in this case. Prove that the two definitions are indeed
+equivalent.
+
+> tr_rev_correct : (x : List a) -> tr_rev x = rev x
+> tr_rev_correct = ?tr_rev_correct_rhs
+
+$\square$
+
+
+=== Propositions and Booleans
+
+We've seen two different ways of encoding logical facts in Idris: with
+_booleans_ (of type \idr{Bool}), and with _propositions_ (of type \idr{Type}).
+
+For instance, to claim that a number \idr{n} is even, we can say either
+
+  - (1) that \idr{evenb n} returns \idr{True}, or
+
+  - (2) that there exists some \idr{k} such that \idr{n = double k}. Indeed,
+    these two notions of evenness are equivalent, as can easily be shown with a
+    couple of auxiliary lemmas.
+
+We often say that the boolean \idr{evenb n} _reflects_ the proposition \idr{(k
+** n = double k)}.
+
+> double : (n : Nat) -> Nat      
+> double  Z    = Z               
+> double (S k) = S (S (double k))
+
+> evenb_double : evenb (double k) = True
+> evenb_double {k = Z} = Refl
+> evenb_double {k = (S k')} = evenb_double {k=k'}
+
+
+==== Exercise: 3 stars (evenb_double_conv)
+
+> evenb_double_conv : (k ** n = if evenb n then double k else S (double k))
+> -- Hint: Use the \idr{evenb_S} lemma from `Induction`
```

Comment:
This should be `` `evenb_S` `` instead of `\idr{evenb_S}`. Alternatively we could do something like this instead:

````markdown
> evenb_double_conv : (k ** n = if evenb n then double k else S (double k))

*Hint*: Use the \idr{evenb_S} lemma from \idr{Induction}.

> evenb_double_conv = ?evenb_double_conv_rhs
````